apps/vault.go

```diff
@@ -1,10 +1,251 @@
+}
+
+const (
+	defaultVaultEndpoint = "localhost:8200"
+	defaultMetricsPath   = "/v1/sys/metrics"
+)
+
+func (r MetricsReceiverVault) Type() string {
+	return "vault"
+}
+
+func (r MetricsReceiverVault) Pipelines() []otel.Pipeline {
+	if r.Endpoint == "" {
+		r.Endpoint = defaultVaultEndpoint
+	}
+	if r.MetricsPath == "" {
+		r.MetricsPath = defaultMetricsPath
+	}
+
+	tlsConfig := r.TLSConfig(true)
+
+	scrapeConfig := map[string]interface{}{
+		"job_name":        "vault",
+		"scrape_interval": r.CollectionIntervalString(),
+		"metrics_path":    r.MetricsPath,
+		"static_configs": []map[string]interface{}{{
+			"targets": []string{r.Endpoint},
+		}},
+	}
+
+	if r.Token != "" {
+		scrapeConfig["scheme"] = "https"
+		scrapeConfig["authorization"] = map[string]interface{}{
+			"credentials": r.Token,
+			"type":        "Bearer",
+		}
+	}
+	if tlsConfig["insecure"] == false {
+		scrapeConfig["scheme"] = "https"
+	}
+	delete(tlsConfig, "insecure")
+	scrapeConfig["tls_config"] = tlsConfig
+
+	includeMetrics := []string{}
+
+	storageMetricTransforms, newStorageMetricNames := r.addStorageMetrics()
+	metricDetailTransforms, newMetricNames := r.getMetricTransforms()
+
+	includeMetrics = append(includeMetrics, newStorageMetricNames...)
+	includeMetrics = append(includeMetrics, newMetricNames...)
+
+	return []otel.Pipeline{{
+		Receiver: otel.Component{
+			Type: "prometheus",
+			Config: map[string]interface{}{
+				"config": map[string]interface{}{
+					"scrape_configs": []map[string]interface{}{
+						scrapeConfig,
+					},
+				},
+			},
+		},
+		Processors: []otel.Component{
+			storageMetricTransforms,
+			metricDetailTransforms,
+			otel.MetricsFilter(
+				"include",
+				"strict",
+				includeMetrics...,
+			),
+			otel.NormalizeSums(),
+			otel.MetricsTransform(
+				otel.AddPrefix("workload.googleapis.com"),
+			),
+		},
+	}}
+}
+
+type metricTransformer struct {
+	OldName     string
+	NewName     string
+	Description string
+	Unit        string
+}
+
+func (r MetricsReceiverVault) addStorageMetrics() (transforms otel.Component, newMetrics []string) {
+	storageLabel := "storage"
+
+	storages := []string{
+		"zookeeper",
+		"swift",
+		"spanner",
+		"s3",
+		"postgres",
+		"mysql",
+		"mssql",
+		"gcs",
+		"etcd",
+		"dynamodb",
+		"couchdb",
+		"consul",
+		"cockroachdb",
+		"cassandra",
+		"azure",
+	}
+
+	operations := []string{
+		"put",
+		"get",
+		"delete",
+		"list",
+	}
+
+	queries := []otel.TransformQuery{}
+
+	for _, operation := range operations {
+		operationTimeName := "vault.storage.operation." + operation + ".time"
+		operationCountName := "vault.storage.operation." + operation + ".count"
+		newMetrics = append(newMetrics, operationTimeName)
+		newMetrics = append(newMetrics, operationCountName)
+		for _, storage := range storages {
+			oldMetric := "vault_" + storage + "_" + operation
+
+			queries = append(queries, otel.SummaryCountValToSum(oldMetric, "cumulative", true))
```

Comment:
Should we be dropping the summary metric after we've gotten the parts of it that we want separated out?

Reply:
it gets filtered out because it's specified to only keep new metrics